### PR TITLE
get byUTF() to work with -dip1000

### PR DESCRIPTION
**Pull request by WalterBright**
The trouble here is `decodeImpl()` uses pointers directly for speed, meaning the function is not `@safe`. Being not `@safe` means it does not infer `scope` for its parameters, and then `popFront()` doesn't infer `scope` either, causing cascading problems up the call stack.

To fix is a bit clumsy, saying that `popFront()` is `scope`, and then trusting `decodeImpl()`. Note that inferring `@safe` for `popFront()` still works correctly, since it winds up calling `empty` and `front` and `popFront` so if those aren't `@safe` then it won't be inferred `@safe`.

P.S. `scope` cannot be inferred reliably for `@system` code.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
any performance implications by called another delegate ?

**Comment by WalterBright** [comment]
It should be inlined by the compiler.

**Comment by WalterBright** [comment]
Looks like dscanner in ci/circleci has a problem with:
```
../dscanner/dsc --config .dscanner.ini --styleCheck $(find etc std -type f -name '*.d' | grep -vE 'std/traits.d|std/typecons.d') -I.
std/utf.d(3810:40)[error]: 'in', 'out', 'body', or block statement expected
```
ont the line:
```
@property auto front() scope
```
I think it is failing on the `scope`. dscanner needs updating.

**Comment by JackStouffer** [comment]
Wouldn't it just make more sense to remove the `const char[]` branch from `decodeImpl` and use the code from the `hasSlicing` branch for strings?

**Comment by jmdavis** [comment]
> Wouldn't it just make more sense to remove the const char[] branch from decodeImpl and use the code from the hasSlicing branch for strings?

That would incur additional overhead, because then the string would be bounds checked on every access, and the logic of `decodeImpl` already takes care of the bounds checking and either throws a `UTFException` or uses `replacementDchar` when the range isn't long enough.

File: std/utf.d [hunk 1]
```diff
@@ -3829,14 +3824,12 @@ template byUTF(C) if (isSomeChar!C)
                         {
                             static if (is(RC == dchar))
                             {
-                                fill = cast(ushort) encode!(Yes.useReplacementDchar)(buf, c);
                                 r.popFront;
+                                dchar dc = c;
                             }
                             else
-                            {
```
Comment (JackStouffer) [inline std/utf.d]: style nitpick: If one part of the static if uses braces, all of them should use it and vice versa.

File: std/utf.d [hunk 2]
```diff
@@ -3802,17 +3802,12 @@ template byUTF(C) if (isSomeChar!C)
...
-                    this.r = r;
-                }
-
                 @property bool empty()
                 {
                     return pos == fill && r.empty;
                 }
 
-                @property auto front()
+                @property auto front() scope // 'scope' required by call to decodeFront() below
```
Comment (JackStouffer) [inline std/utf.d]: So I'm trying to read and understand the DIP, please correct me on any details I have wrong:

Adding scope here adds checks to the `this` reference to the struct itself to make sure references to member variables aren't escaped from the function calls and return values. This does not change the return value of the function and this change should be completely transparent to current code.

File: std/utf.d (hunk 1 above)
Comment (jmdavis) [inline std/utf.d]: Personally, I see no problem with having braces on one branch and not on another. If the branch is only one line, then braces are just noise. But the style guide doesn't say anything on the matter.

File: std/utf.d (hunk 2 above)
Comment (WalterBright) [inline std/utf.d]: yes.

**Comment by JackStouffer** [comment]
Auto-merge toggled on

**Pull request merged**